README.md
# Gitapp
# HOLIS

Aplicación utilizada en la cursada 2021 de Ingenieria de Software. UTN-FRLP

## Dependencias

-   node > v14
-   express > v4

## Instalar dependencias

`npm install`

## Iniciar la app en modo desarrollo

`npm run dev`

## Iniciar la app en modo producción

`npm start`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2021-06-17

### Added

-   Agregar validación de HTML5 en el form de crear movimiento
-   Crear vista de egresos de manera análoga a la tabla de ingresos
-   Agregar campo description al modelo
-   Agregar un signo “-” delante de cada gasto y un signo “+” delante de cada ingreso de
todas las listas

## [1.1.0] - 2021-06-16

### Added

-   Alerta que indica que un movimiento se guardó exitosamente
-   Agregar endpoint en la API para permitir borrar un movimiento

### Fixed

-   Agregacion de los headers de la card de los gráficos
-   Correcion de fecha en creacion de movimientos
-   Arreglar el formato de los montos en las tablas

## [1.0.1] - 2021-05-03

### Added

-   Cypress detection for running tests on memory
-   Cypress seed before each cypress test

### Changed

-   Creates tables on server init and avoids erase on shutdown

### Removed

-   Cypress experimental configuration

## [1.0.0] - 2021-04-26

### Added

-   Movements API
-   Home UI with charts and last movements
-   Incomes UI with last incomes

[unreleased]: https://github.com/lucasantarelli18/gitapp/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/lucasantarelli18/gitapp/releases/tag/v1.2.0
[1.1.0]: https://github.com/lucasantarelli18/gitapp/releases/tag/v1.1.0
[1.0.1]: https://github.com/lucasantarelli18/gitapp/releases/tag/v1.0.1
[1.0.0]: https://github.com/lucasantarelli18/gitapp/releases/tag/v1.0.0


tests/unit/ui.spec.js
const render = require('../nunjucks-util.js');

const { getByText, getByTestId } = require('@testing-library/dom');
require('@testing-library/jest-dom');

test('Deberia renderizar correctamente un movimiento del tipo income', () => {
    const $movement = render('movement', {
        movement: {
            type: 'income',
            date: '01/01/2021',
            amount: 1000.0,
            category: 'Supermercado',
        },
    });

    getByText($movement, 'Supermercado');
    expect(getByTestId($movement, 'movement-icon')).toHaveAttribute(
        'src',
        expect.stringContaining('income')
    );
    expect(getByText($movement, '+ $')).toHaveClass('has-text-success');
});


test('Deberia figurar un + adelante de los ingresos de las listas', () => {
    const $movement = render('movement', {
        movement: {
            type: 'income',
            date: '01/01/2021',
            amount: 1000.0,
            category: 'Supermercado',
        },
    });

    expect(getByText($movement, '+ $'));
});

test('Deberia figurar un - adelante de los ingresos de las listas', () => {
    const $movement = render('movement', {
        movement: {
            type: 'expense',
            date: '01/01/2021',
            amount: 1000.0,
            category: 'Supermercado',
        },
    });

    expect(getByText($movement, '- $'));
});



tests/unit/utils.spec.js
const { formatDate } = require('../../client/js/utils.js');
const { monefy } = require('../../client/js/utils.js');

describe('Método formatDate', () =>{

    const fecha = new Date('August 20, 2020 23:15:33');
    const fechaEsperada = '20/8/2020';
    it('Dada una fecha, debería devolverla en el formato correcto', () => {
        expect(formatDate(fecha)).toBe(fechaEsperada);
    });


});

describe('Método monefy', () => {

    it('Debería convertir 1000.35 en "1000,35"', async () => {
        expect(monefy(1000.35)).toBe('1000,35');
    });
    
    it('Debería convertir 100 en "100"', async () => {
        expect(monefy(100)).toBe('100');
    });
    
    it('Debería convertir undefined en ""', async () => {
        expect(monefy(undefined)).toBe('');
    });


});

tests/unit/models.spec.js
const MovementModel = require('../../server/models/movement.js');
const MovementType = require('../../server/models/movementType.js');

beforeEach(async () => {
    await MovementModel.Movement.sync({ force: true });
});

test('Crear movimiento', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    expect(movement.amount).toBe(movementData.amount);
    expect(movement.type).toBe(movementData.type);
    expect(movement.category).toBe(movementData.category);
});

test('Tomar la fecha enviada', async () => {
    const movementData = {
        date: '2021-04-01T03:00:00.000Z',
        amount: 50000.0,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    expect(Date.parse(movement.date)).toBe(Date.parse(movementData.date));
});

test('Crear movimiento sin tipo', async () => {
    const movementData = {
        date: '01/01/2021',
        amount: 1000.0,
        category: 'Supermercado',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    expect(movement.amount).toBe(movementData.amount);
    expect(movement.type).toBe(MovementType.EXPENSE);
    expect(movement.category).toBe(movementData.category);
});

test('Crear movimiento sin fecha', async () => {
    const movementData = {
        amount: 1000.0,
        category: 'Supermercado',
    };

    try {
        await MovementModel.create(movementData);
    } catch (e) {
        expect(e.name).toBe('SequelizeDatabaseError');
    }
});

test('Listar movimientos sin resultados', async () => {
    const movements = await MovementModel.getAll();

    expect(movements).not.toBeNull();
    expect(movements.rows.length).toBe(0);
});

test('Listar movimientos con resultados', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    await MovementModel.create(movementData);

    const movements = await MovementModel.getAll();

    expect(movements).not.toBeNull();
    expect(movements.rows.length).toBe(1);
});

test('Listar movimientos con limite', async () => {
    const firstMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    const secondMovementData = {
        date: '05/01/2021',
        amount: 1000.0,
        category: 'Supermercado',
    };

    // Creamos los movimientos
    const movement = await MovementModel.create(firstMovementData);
    await MovementModel.create(secondMovementData);

    let movements = await MovementModel.getAll(1);

    // La lista de movimientos debería contener solo el primero
    expect(movements.rows.length).toBe(1);
    expect(movements.rows[0].id).toBe(movement.id);
});

test('Listar movimientos con limite y offset', async () => {
    const firstMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    const secondMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos los movimientos
    await MovementModel.create(firstMovementData);
    const movement = await MovementModel.create(secondMovementData);

    let movements = await MovementModel.getAll(1, 1);

    // La lista de movimientos debería contener solo el segundo
    expect(movements.rows.length).toBe(1);
    expect(movements.rows[0].id).toBe(movement.id);
});

test('Filtrar movimientos por tipo income', async () => {
    const firstMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    const secondMovementData = {
        date: '05/01/2021',
        amount: 1000.0,
        category: 'Supermercado',
    };

    // Creamos los movimientos
    const movement = await MovementModel.create(firstMovementData);
    await MovementModel.create(secondMovementData);

    let movements = await MovementModel.getAll(null, null, MovementType.INCOME);

    // La lista de movimientos debería contener solo el primero
    expect(movements.rows.length).toBe(1);
    expect(movements.rows[0].id).toBe(movement.id);
});

test('Filtrar movimientos por tipo expense', async () => {
    const firstMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    const secondMovementData = {
        date: '05/01/2021',
        amount: 1000.0,
        category: 'Supermercado',
    };

    // Creamos los movimientos
    await MovementModel.create(firstMovementData);
    const movement = await MovementModel.create(secondMovementData);

    let movements = await MovementModel.getAll(
        null,
        null,
        MovementType.EXPENSE
    );

    // La lista de movimientos debería contener solo el segundo
    expect(movements.rows.length).toBe(1);
    expect(movements.rows[0].id).toBe(movement.id);
});

test('Filtrar movimientos por tipo inexistente', async () => {
    const firstMovementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    const secondMovementData = {
        date: '05/01/2021',
        amount: 1000.0,
        category: 'Supermercado',
    };

    // Creamos los movimientos
    await MovementModel.create(firstMovementData);
    await MovementModel.create(secondMovementData);

    let movements = await MovementModel.getAll(null, null, 'fake');

    // La lista de movimientos debería estar vacía
    expect(movements.rows.length).toBe(0);
});

test('Editar movimiento', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    // La categoría debería ser la creada
    expect(movement.category).toBe(movementData.category);

    const updateData = {
        category: 'Otros',
    };

    // Modificamos categoría del movimiento
    const movementUpdated = await MovementModel.update(movement.id, updateData);

    // La función debería retornar algo
    expect(movementUpdated).not.toBeNull();

    // La categoría debería ser la modificada
    expect(movementUpdated.category).toBe('Otros');
});

test('Editar movimiento inexistente', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    // La categoría debería ser la creada
    expect(movement.category).toBe(movementData.category);

    const updateData = {
        category: 'Otros',
    };

    // Modificamos movimiento inexistente
    const movementUpdated = await MovementModel.update(2, updateData);

    // La función debería retornar null
    expect(movementUpdated).toBeNull();

    const movements = await MovementModel.getAll();

    // La categoría debería seguir siendo la misma
    expect(movements.rows[0].category).toBe(movementData.category);
});

test('Eliminar movimiento', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos el movimiento
    const movement = await MovementModel.create(movementData);

    // Buscamos todos los movimientos
    let movements = await MovementModel.getAll();

    // Debe existir un movimiento en la lista
    expect(movements.rows.length).toBe(1);

    // Eliminamos movimiento
    const deleted = await MovementModel.delete(movement.id);

    // La función debería retornar algo
    expect(deleted).not.toBeNull();

    movements = await MovementModel.getAll();

    // No deben haber movimientos en la lista
    expect(movements.rows.length).toBe(0);
});

test('Eliminar movimiento inexistente', async () => {
    const movementData = {
        date: '04/01/2021',
        amount: 50000.0,
        type: MovementType.INCOME,
        category: 'Sueldo',
    };

    // Creamos un movimiento
    await MovementModel.create(movementData);

    // Buscamos todos los movimientos
    let movements = await MovementModel.getAll();

    // Debe existir un movimiento en la lista
    expect(movements.rows.length).toBe(1);

    // Eliminamos movimiento inexistente
    const deleted = await MovementModel.delete(2);

    // La función debería retornar null
    expect(deleted).toBeNull();

    movements = await MovementModel.getAll();

    // El movimiento debería seguir existiendo en la lista
    expect(movements.rows.length).toBe(1);
});


client/js/utils.js
export function monefy(num) {
    if (!num) return '';

    const NumS = String(num);

    const Total = NumS.split('.');


    const numStr = String(Total[0]);


    const points = numStr.length / 3;

    const result = [];

    for (let i = 0; i < points; i++) {
        const s = -3 * (i + 1);
        const e = -3 * i || undefined;
        const chunk = numStr.slice(s, e);


        result.push(chunk);
    }

    if (Total[1] == null){
        return result.reverse().join('.');
    } else{
        return result.reverse().join('.') + ',' + Total[1];
    }
    
}

export function getRandomColor() {
    return `hsl(${360 * Math.random()}, ${25 + 70 * Math.random()}%, ${
        65 + 10 * Math.random()
    }%)`;
}

export function getMonth(dateString) {
    const monthNames = [
        'Enero',
        'Febrero',
        'Marzo',
        'Abril',
        'Mayo',
        'Junio',
        'Julio',
        'Agosto',
        'Septiembre',
        'Octubre',
        'Noviembre',
        'Diciembre',
    ];

    const date = new Date(dateString);

    return monthNames[date.getMonth()];
}

export function formatDate(date) {
    return date.split('T')[0];
}


client/js/income.js
import { getRefs, render } from './render.js';
import movementService from './movement-service.js';

let state = {
    movements: [],
    movement: {},
    hasEdit: true,
};

let refs = getRefs(document.body);

/**
 * Obtiene todos los ultimos movimientos disponibles
 **/
async function getIncomes() {
    return movementService.getIncomes();
}

/**
 * Renderiza los libros
 **/
function renderIncomes(state) {
    render('movement-list.html', state, refs.incomes);
}

/**
 * Inicializa la vista income
 **/
async function init() {
    state.movements = await getIncomes();
    renderIncomes(state);
}

function getMovementData() {
    const formData = new FormData(refs.form.firstElementChild);
    const movement = Object.fromEntries(formData);
    movement.type = "income"
    return movement;
}

// Event Listeners

/**
 * Agrega un movimiento a edicion
 **/
window.editMovement = function (movement) {
    state.movement = movement;
    render('movement-form.html', state, refs.form);
};

/**
 * Cancela una edicion o creación
 **/
window.onCancel = function () {
    state.movement = {};
    render('movement-form.html', state, refs.form);
};

/**
 * Elimina un movimiento
 **/
window.onRemove = async function () {
    await movementService.update(state.movement);
    state.movement = {};
    render('movement-form.html', state, refs.form);
};

/**
 * Guarda un movimiento
 **/
window.onSave = async function (e) {

    e.stopPropagation();

    const form = document.querySelector('form');

    if (!form.checkValidity()){
        return ;
    }

    e.preventDefault();
    const movement = getMovementData();

    if (movement.id) {
        await movementService.update(movement);
    } else {
        await movementService.create(movement);
    }

    state.movement = {};
    render('movement-form.html', state, refs.form);
};

init();


client/js/expense.js
import { getRefs, render } from './render.js';
import movementService from './movement-service.js';

let state = {
    movements: [],
    movement: {},
    hasEdit: true,
};

let refs = getRefs(document.body);

/**
 * Obtiene todos los ultimos movimientos disponibles
 **/
async function getExpense() {
    return movementService.getExpense();
}

/**
 * Renderiza los libros
 **/
function renderExpense(state) {
    render('movement-list.html', state, refs.expense);
}

/**
 * Inicializa la vista expense
 **/
async function init() {
    state.movements = await getExpense();
    renderExpense(state);
}

function getMovementData() {
    const formData = new FormData(refs.form.firstElementChild);
    const movement = Object.fromEntries(formData);
    movement.type = "expense"
    return movement;
}

// Event Listeners

/**
 * Agrega un movimiento a edicion
 **/
window.editMovement = function (movement) {
    state.movement = movement;
    render('movement-form.html', state, refs.form);
};

/**
 * Cancela una edicion o creación
 **/
window.onCancel = function () {
    state.movement = {};
    render('movement-form.html', state, refs.form);
};

/**
 * Elimina un movimiento
 **/
window.onRemove = async function () {
    await movementService.remove(state.movement);
    state.movement = {};
    render('movement-form.html', state, refs.form);
};

/**
 * Guarda un movimiento
 **/
window.onSave = async function (e) {
    e.stopPropagation();
    e.preventDefault();
    const movement = getMovementData();

    if (movement.id) {
        await movementService.update(movement);
    } else {
        await movementService.create(movement);
    }

    state.movement = {};
    render('movement-form.html', state, refs.form);
};

init();


server/routes/movement.js
const express = require('express');
const paginate = require('express-paginate');
const MovementModel = require('../models/movement');
const router = express.Router();

/**
 * Endpoint para obtener todos los movimientos.
 *
 */
router.get('/', function (req, res) {
    MovementModel.getAll(req.query.limit, req.skip, req.query.type)
        .then((results) => {
            const pageCount = Math.ceil(results.count / req.query.limit);

            const pages = paginate.getArrayPages(req)(
                3,
                pageCount,
                req.query.page
            );

            res.status(200).send({
                movements: results.rows,
                pages: pages,
                next: paginate.hasNextPages(req)(pageCount)
                    ? pages[req.query.page].url
                    : null,
            });
        })
        .catch((_) => {
            console.log(_);
            res.status(500).send('Error al obtener los movimientos');
        });
});

/**
 * Endpoint para crear un movimiento.
 * Recibe los datos del movimiento en body.
 *
 */
router.post('/', function (req, res) {
    MovementModel.create(req.body)
        .then((data) => {
            res.status(201).send(data);
        })
        .catch((_) => {
            console.log(_);
            res.status(500).send('Error al crear movimiento');
        })
});

/**
 * Endpoint para editar un movimiento.
 * Recibe el id en req.params.id
 *
 */
router.put('/:id', function (req, res) {
    MovementModel.update(req.params.id, req.body)
        .then((movement) => {
            if (movement == null) {
                res.status(404).send(
                    'El movimiento ' + req.params.id + ' no fue encontrado'
                );
            } else res.status(200).send(movement);
        })
        .catch(() => res.status(500).send('Error al obtener movimiento'));
});


/**
 * Endpoint para eliminar un movimiento.
 * Recibe el id en req.params.id
 *
 */
router.delete('/:id', function (req, res) {
    MovementModel.delete(req.params.id)
        .then((movement) => {
            if (movement == null) {
                res.status(404).send(
                    'El movimiento ' + req.params.id + ' no fue encontrado'
                );
            } else res.status(200).send(movement);
        })
        .catch(() => res.status(500).send('Error al obtener movimiento'));
});


module.exports = router;


client/js/movement-service.js
const BASE_URL = '/api/v1';

async function getLast() {
    const resp = await fetch(`${BASE_URL}/movements`);
    const { movements } = await resp.json();
    return movements;
}

async function getIncomes() {
    const resp = await fetch(`${BASE_URL}/movements?type=income`);
    const { movements } = await resp.json();
    return movements;
}

async function getExpense() {
    const resp = await fetch(`${BASE_URL}/movements?type=expense`);
    const { movements } = await resp.json();
    return movements;
}

async function update(movement) {
    const resp = await fetch(`${BASE_URL}/movements/${movement.id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(movement),
    });

    return resp.json();
}

async function create(movement) {
    const resp = await fetch(`${BASE_URL}/movements`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(movement),
        
    });

    window.alert("El movimiento se guardo con exito");

    return resp.json();

}

async function remove(movement) {
    const resp = await fetch(`${BASE_URL}/movements/${movement.id}`, {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(movement),
    });

    console.log('delete:', movement);

    return new Promise(resolve => {
        resolve();
        resp.json();
    });
}

export default {
    create,
    update,
    remove,
    getLast,
    getIncomes,
    getExpense,
};


scripts/fixture.js
const { Movement } = require('../server/models/movement.js');
const MovementType = require('../server/models/movementType.js');

const movements = [
    {
        date: '01/01/2021',
        amount: 1000,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '01/04/2021',
        amount: 587.5,
        type: MovementType.EXPENSE,
        category: 'Librería',
    },
    {
        date: '01/04/2021',
        amount: 50000,
        type: MovementType.INCOME,
        category: 'Sueldo',
    },
    {
        date: '01/04/2021',
        amount: 233,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '01/05/2021',
        amount: 10000,
        type: MovementType.INCOME,
        category: 'Plazo Fijo',
    },
    {
        date: '01/07/2021',
        amount: 847,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '01/08/2021',
        amount: 751,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '01/10/2021',
        amount: 2500,
        type: MovementType.EXPENSE,
        category: 'Reparación Celular',
    },
    {
        date: '01/14/2021',
        amount: 1432,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '01/15/2021',
        amount: 2300,
        type: MovementType.EXPENSE,
        category: 'Librería',
    },
    {
        date: '01/20/2021',
        amount: 4877,
        type: MovementType.EXPENSE,
        category: 'Regalos',
    },
    {
        date: '01/21/2021',
        amount: 2900,
        type: MovementType.EXPENSE,
        category: 'Regalos',
    },
    {
        date: '02/04/2021',
        amount: 50000,
        type: MovementType.INCOME,
        category: 'Sueldo',
    },
    {
        date: '02/05/2021',
        amount: 11000,
        type: MovementType.INCOME,
        category: 'Plazo Fijo',
    },
    {
        date: '02/07/2021',
        amount: 2100,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '02/08/2021',
        amount: 876,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '02/10/2021',
        amount: 1398,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '02/15/2021',
        amount: 3456,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
    {
        date: '02/20/2021',
        amount: 1498,
        type: MovementType.EXPENSE,
        category: 'Supermercado',
    },
];

const init = () => {
    return Movement.sync({ force: true }).then(() =>
        Movement.bulkCreate(movements)
    );
};

if (require.main === module) {
    init();
}

module.exports = {
    init,
};
